src/Processors/QueryPlan/Optimizations/optimizeUseProjections.cpp

```diff
@@ -0,0 +1,1192 @@
+
+    return nullptr;
+}
+
+static bool canUseProjectionForReadingStep(ReadFromMergeTree * reading)
+{
+    /// Probably some projection already was applied.
+    if (reading->hasAnalyzedResult())
+        return false;
+
+    if (reading->isQueryWithFinal())
+        return false;
+
+    if (reading->isQueryWithSampling())
+        return false;
+
+    if (reading->isParallelReadingEnabled())
+        return false;
+
+    // Currently projection don't support deduplication when moving parts between shards.
+    if (reading->getContext()->getSettingsRef().allow_experimental_query_deduplication)
+        return false;
+
+    return true;
+}
+
+bool optimizeUseAggProjections(QueryPlan::Node & node, QueryPlan::Nodes & nodes)
+{
+    if (node.children.size() != 1)
+        return false;
+
+    auto * aggregating = typeid_cast<AggregatingStep *>(node.step.get());
+    if (!aggregating)
+        return false;
+
+    if (!aggregating->canUseProjection())
+        return false;
+
+    QueryPlan::Node * reading_node = findReadingStep(*node.children.front());
+    if (!reading_node)
+        return false;
+
+    auto * reading = typeid_cast<ReadFromMergeTree *>(reading_node->step.get());
+    if (!reading)
+        return false;
+
+    if (!canUseProjectionForReadingStep(reading))
+        return false;
+
+    std::shared_ptr<PartitionIdToMaxBlock> max_added_blocks = getMaxAddedBlocks(reading);
+
+    auto candidates = getAggregateProjectionCandidates(node, *aggregating, *reading, max_added_blocks);
+
+    AggregateProjectionCandidate * best_candidate = nullptr;
+    if (candidates.minmax_projection)
+        best_candidate = &candidates.minmax_projection->candidate;
+    else if (candidates.real.empty())
+        return false;
+
+    const auto & parts = reading->getParts();
+    const auto & query_info = reading->getQueryInfo();
+    const auto metadata = reading->getStorageMetadata();
+    ContextPtr context = reading->getContext();
+    MergeTreeDataSelectExecutor reader(reading->getMergeTreeData());
+
+    /// Selecting best candidate.
+    for (auto & candidate : candidates.real)
+    {
+        auto required_column_names = candidate.dag->getRequiredColumnsNames();
+        ActionDAGNodes added_filter_nodes;
+        if (candidates.has_filter)
+            added_filter_nodes.nodes.push_back(candidate.dag->getOutputs().front());
+
+        bool analyzed = analyzeProjectionCandidate(
+            candidate, *reading, reader, required_column_names, parts,
+            metadata, query_info, context, max_added_blocks, added_filter_nodes);
+
+        if (!analyzed)
+            continue;
+
+        if (best_candidate == nullptr || best_candidate->sum_marks > candidate.sum_marks)
+            best_candidate = &candidate;
+    }
+
+    if (!best_candidate)
+        return false;
+
+    QueryPlanStepPtr projection_reading;
+    bool has_nornal_parts;
+
+    /// Add reading from projection step.
+    if (candidates.minmax_projection)
```

Comment:
Quite a tricky place to read. If it's not null, then `best_candidate` is set to `candidates.minmax_projection`.